### ISSUE TITLE
Custom lifecycle hooks configured via CDIConfig

### ISSUE THREAD

**Issue by bc185174** [kubevirt/containerized-data-importer#3012]
**Is your feature request related to a problem? Please describe**:
A clear and concise description of what the problem is.

In https://github.com/kubevirt/containerized-data-importer/issues/2948 we found that the CDIs transient pods (k8s jobs) were not terminated when meshed with linkerd. The current workaround is to disable injection on these pods by propagating the annotation `linkerd.io/inject: disabled` from the `DataVolume` to the importer pods. Ultimatly, we would like linkerd to implement sidecars through native k8s functionality; however, this is only available in `v1.28.x` and users will be unlikely to take alpha version of security mesh provider.

**Describe the solution you'd like**:

Ability to add custom lifecycle hooks to the importer pods via `CDIConfig`.

```
lifecycle:
  preStop:
    exec:
       command:
       - /bin/bash
       - -c
       - curl -X POST http://localhost:4191/shutdown
```

**Describe alternatives you've considered**:

N/A

**Additional context**:

**Comment by bc185174** [comment]
fyi @akalenyu

**Comment by awels** [comment]
Would it make sense to use something like [Apache airflow](https://airflow.apache.org/docs/apache-airflow/2.0.0/kubernetes.html) to generate a mutation webhook that injects that into the importer pods?

**Comment by bc185174** [comment]
> Would it make sense to use something like [Apache airflow](https://airflow.apache.org/docs/apache-airflow/2.0.0/kubernetes.html) to generate a mutation webhook that injects that into the importer pods?

Not sure it would - we don't really want to deploy another control-plane and add extra load to handle this case.

**Comment by bc185174** [comment]
I can try get a PR up soon to contribute this change.

**Comment by akalenyu** [comment]
> I can try get a PR up soon to contribute this change.

We actually discussed this in one of the sig-storage meetings, let me find the summary

**Comment by bc185174** [comment]
> > I can try get a PR up soon to contribute this change.
> 
> We actually discussed this in one of the sig-storage meetings, let me find the summary

@akalenyu any update on this? 

I've also just hit the same issue with transient VM pods (shutting down vms). What I'd propose is to assign that lifecycle hook if the annotation `linkerd.io/inject: enabled` is preset as opposed to making it configurable.

**Comment by akalenyu** [comment]
> > > I can try get a PR up soon to contribute this change.
> > 
> > 
> > We actually discussed this in one of the sig-storage meetings, let me find the summary
> 
> @akalenyu any update on this?
> 
> I've also just hit the same issue with transient VM pods (shutting down vms). What I'd propose is to assign that lifecycle hook if the annotation `linkerd.io/inject: enabled` is preset as opposed to making it configurable.

Hey sorry forgot to attach the link:
https://youtu.be/5YoeDestQpw?t=2081
TL DR there wasn't much appetite since the feature was going to land in 1.28. Is this not the case anymore?

**Comment by akalenyu** [comment]
> I've also just hit the same issue with transient VM pods (shutting down vms). What I'd propose is to assign that lifecycle hook if the annotation `linkerd.io/inject: enabled` is preset as opposed to making it configurable.

TBH doesn't sound too bad. Are there instances where one would *NOT* want to issue this shutdown on exit?

**Comment by akalenyu** [comment]
Summary from sig-storage:
We are totally okay with introducing this annotation to send the shutdown signal.
Might as well add support for istio as well (kubevirt/kubevirt already has this implemented) if we're doing this

/assign bc185174

**Comment by bc185174** [comment]
Going to close this for now. Adding the lifecycle hook does not resolve the issue since its not triggered if the container is shutdown gracefully. Will look at what other approaches to take.

**Comment by bc185174** [comment]
Yep happy to close, thank you!
